packages/beacon-node/src/chain/blocks/types.ts

```diff
@@ -32,26 +32,121 @@ export function blockRequiresBlobs(config: ChainForkConfig, blockSlot: Slot, clo
+  | {type: GossipedInputType.blob; signedBlob: deneb.SignedBlobSidecar; blobBytes: Uint8Array | null};
+type BlockInputCacheType = {
+  block?: allForks.SignedBeaconBlock;
+  blockBytes?: Uint8Array | null;
+  blobs: Map<number, deneb.BlobSidecar>;
+  blobsBytes: Map<number, Uint8Array | null>;
+};
+
+const MAX_GOSSIPINPUT_CACHE = 5;
+// TODO deneb: export from types package
+// ssz.deneb.BlobSidecars.elementType.fixedSize;
+const BLOBSIDECAR_FIXED_SIZE = 131256;
 
 export const getBlockInput = {
+  blockInputCache: new Map<RootHex, BlockInputCacheType>(),
+
+  getGossipBlockInput(
+    config: ChainForkConfig,
+    gossipedInput: GossipedBlockInput
+  ):
+    | {blockInput: BlockInput; blockInputMeta: {pending: null; haveBlobs: number; expectedBlobs: number}}
+    | {blockInput: null; blockInputMeta: {pending: GossipedInputType.block; haveBlobs: number; expectedBlobs: null}}
+    | {blockInput: null; blockInputMeta: {pending: GossipedInputType.blob; haveBlobs: number; expectedBlobs: number}} {
+    let blockHex;
+    let blockCache;
+
+    if (gossipedInput.type === GossipedInputType.block) {
+      const {signedBlock, blockBytes} = gossipedInput;
+
+      blockHex = toHexString(
+        config.getForkTypes(signedBlock.message.slot).BeaconBlock.hashTreeRoot(signedBlock.message)
+      );
+      blockCache = this.blockInputCache.get(blockHex) ?? {
+        blobs: new Map<number, deneb.BlobSidecar>(),
+        blobsBytes: new Map<number, Uint8Array | null>(),
+      };
+
+      blockCache.block = signedBlock;
+      blockCache.blockBytes = blockBytes;
+    } else {
+      const {signedBlob, blobBytes} = gossipedInput;
+      blockHex = toHexString(signedBlob.message.blockRoot);
+      blockCache = this.blockInputCache.get(blockHex);
+
+      // If a new entry is going to be inserted, prune out old ones
+      if (blockCache === undefined) {
+        pruneSetToMax(this.blockInputCache, MAX_GOSSIPINPUT_CACHE);
+        blockCache = {blobs: new Map<number, deneb.BlobSidecar>(), blobsBytes: new Map<number, Uint8Array | null>()};
+      }
+
+      // TODO: freetheblobs check if its the same blob or a duplicate and throw/take actions
+      blockCache.blobs.set(signedBlob.message.index, signedBlob.message);
+      // easily splice out the unsigned message as blob is a fixed length type
+      blockCache.blobsBytes.set(signedBlob.message.index, blobBytes?.slice(0, BLOBSIDECAR_FIXED_SIZE) ?? null);
+    }
+
+    this.blockInputCache.set(blockHex, blockCache);
+    const {block: signedBlock, blockBytes} = blockCache;
+
+    if (signedBlock !== undefined) {
+      // block is available, check if all blobs have shown up
+      const {blobKzgCommitments} = (signedBlock as deneb.SignedBeaconBlock).message.body;
+      if (blobKzgCommitments.length < blockCache.blobs.size) {
+        throw Error(`Received more blobs=${blockCache.blobs.size} than commitments=${blobKzgCommitments.length}`);
```

Comment:
we may want to also include block hex + slot here

Reply:
done